### ISSUE TITLE
Spurious counterexample using llvm_verify

### ISSUE THREAD

**Issue by robdockins** [GaloisInc/saw-script#133]
Consider the following simple LLVM module `test.ll`

```
define i64 @basic_add( i64 %x, i64 %y ) {
  %z = add i64 %x, %y
  ret i64 %z
}
```

We attempt to prove this is equivalent to the standard Cryptol addition function with the following script:

```
llvm_module <- llvm_load_module "test.bc";

set_base 16;
print "Proving equivalance";
llvm_verify llvm_module "basic_add" [] do {
  in0 <- llvm_var "x" (llvm_int 64);
  in1 <- llvm_var "y" (llvm_int 64);
  let {{ z = in0 + in1 }};
  llvm_return {{ z }};
  llvm_verify_tactic z3;
};
```

I get the following output:

```
$ clang -c -emit-llvm -o test.bc test.ll
warning: overriding the module target triple with x86_64-apple-macosx10.11.0 [-Woverride-module]
1 warning generated.
$ saw test.saw
Loading module Cryptol
Loading file "test.saw"
Proving equivalance
WARNING: No target data layout found; will use defaults.
When verifying @basic_add:
Proof of return value failed.
Counterexample:
  %x: 0xffffffffffffffff
  %y: 0x0000000000000000
return value
Encountered:  0xffffffffffffffff
evalTermF ExtCns unimplemented (x)
```

The problem appears to be related to the use of `let`.  If I inline the definition of `z` directly into the `llvm_return` call, I instead get the expected behavior (the verification succeeds).  A variety of other modifications also make the proof work (e.g., `let z = {{ in0 + in1 }}` )

**Comment by robdockins** [comment]
Side note: I don't think the ``let var = {{ ... }}`` form solves my actual problem because I want to use Cryptol patterns to introduce variables I reference later, e.g.

```
let {{ (x,y,z) = some_fun in0 in1 in2 }};
```

**Comment by atomb** [comment]
I think there's a widespread bug in the LLVM interface that anything in a specification (e.g., used in `llvm_assert_eq`, `llvm_return`, etc.) that's hidden under a named constant (as every top-level Cryptol name is) and refers to the results of `llvm_var` is treated incorrectly. I think I know how to deal with this, fortunately, and hope to get to it soon.

**Comment by atomb** [comment]
I had hoped this would be fixed in the Crucible backend, but it still seems to be an issue. The following SAW file leads to a similar counterexample:

````
load_crucible_llvm_module "test.bc";

set_base 16;
print "Proving equivalance";
crucible_llvm_verify "basic_add" [] false do {
  in0 <- crucible_fresh_var "x" (llvm_int 64);
  in1 <- crucible_fresh_var "y" (llvm_int 64);
  crucible_execute_func [crucible_term in0, crucible_term in1];
  let {{ z = in0 + in1 }};
  crucible_return (crucible_term {{ z }});
}
z3;
````

And, like with `llvm_verify`, putting `in0 + in1` inline in the `crucible_return` statement works.

**Comment by brianhuffman** [comment]
This may have been fixed in 6a74657c146c9ac3244e4e1b7cde270d59855d16 and GaloisInc/saw-core@e0f21245e65d5c1ea42cf3cd5132acf7b59c2bda.
